### PR TITLE
QUIC Frame Encoding and Decoding Functions

### PR DESCRIPTION
**Pull request by hlandau**
Note that this is built on top of the ACKM PR.

I'm holding off on including packet header functions for now until I can ask @mattcaswell if he wants them, since he owns the record layer.

This is done/ready for review so I'm not marking it as a draft, but it shouldn't be merged until ACKM is merged.

### CODE REVIEW

File: test/quic_wire_test.c
Comment (mattcaswell) [inline test/quic_wire_test.c:1046]: `PACKET_remaining()` seems more appropriate.

File: test/quic_wire_test.c
Comment (mattcaswell) [inline test/quic_wire_test.c:65]: I feel like there are perhaps more test cases required here:

E.g. ack frame with no gaps, maximal/minimal ack delay values, no ecn data, etc

File: test/quic_wire_test.c
Comment (mattcaswell) [inline test/quic_wire_test.c:1056]: We don't seem to have any test cases for decoding illegal frame encodings?

File: test/quic_wire_test.c
Comment (mattcaswell) [inline test/quic_wire_test.c:1210]: `PACKET_remaining()`?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Nice code!

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:449]: `has_len` specifies whether the length field is or was included in the encoding. If it isn't, the frame runs to the end of the packet. Maybe could be named less confusingly (`has_explicit_len`?)

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:42]: I agree, but I think it's the most practical choice. The structure would be exactly the same if we redefined it, as in both cases we want a logical representation of the ACK frame contents. There's also a dependency question here; should the wire format PR depend on the ACKM PR or vice versa? If people want the struct defined in this PR and we want to merge this before merging ACKM I can do that. Actually this would have the benefit of making this PR not depend on ACKM. If people are more likely to review this before ACKM, that may be a good idea.

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:193]: Shorter names are always good to me. +1

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:388]: Strictly yes. However we need to advance over the rest of the fields so the PACKET's cursor is at the start of the next frame. "If you pass NULL, the PACKET is in an undefined position on return" is not very helpful (even if in practice I expect users passing NULL will make a copy of the PACKET structure and call it again with the original PACKET structure).

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:586]: This handles both. `expect_frame_header_mask` masks off the LSB, and therefore decodes both types.

There is no output like `int *is_uni` because in order for a user to have called `ossl_quic_wire_decode_max_frame_streams` in the first place, they must already know the frame type, and seeing whether the type is `..._BIDI` or `..._UNI` is trivial for them to do.

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:619]: See above

File: test/quic_wire_test.c
Comment (hlandau) [reply test/quic_wire_test.c:1046]: Good catch

File: test/quic_wire_test.c
Comment (hlandau) [reply test/quic_wire_test.c:65]: The ACK frames are definitely the hairiest frames to encode and decode. There are more tests specifically for ACK below, but more would probably be good.

ACKs with and without ECN are tested, but we can probably test ACK delays more. I'll take this as an action item.

File: test/quic_wire_test.c
Comment (hlandau) [reply test/quic_wire_test.c:1056]: Good point. I'll add some.

File: crypto/packet.c
Comment (kroeckx) [reply crypto/packet.c:210]: I was totally confused, I thought  you changed the len parameter.

File: include/internal/quic_wire.h
Comment (mattcaswell) [reply include/internal/quic_wire.h:449]: I understand what it does. My question really was why we care how it was encoded. Surely the presence or otherwise of the length field is only useful during the decoding itself? i.e. not something we need to pass around and remember after we've decoded it.

File: ssl/quic/quic_wire.c
Comment (mattcaswell) [reply ssl/quic/quic_wire.c:42]: I think it is a bad "smell" indicating that we don't quite understand how the layering is working. What we seem to have here is QUIC wire format encoding/decoding component and an ACK Manager component. The dependency is from the wire format encoder/decoder to the ACK Manager, i.e. the encoder/decoder knows about the ACK Manager, not the other way around.

The design in this PR is the other way around for all the other frame types, i.e. the encoder/decoder component doesn't know anything about any other components, it defines its own structures and expects other components to use them. I think this asymmetry is purely an artefact of the way we have started building the code. We wrote the ACK Manager first, and haven't built other components yet.

One possibility is to move the "ownership" of the ACK structure into the encoder/decoder. That would be most consistent with what you've got for all the other frame types - but feels a bit wrong. In my mind encoding/decoding is a very low-level concept - so having it own the data structures for all the rest of the code feels off.

Another way to design this is to not have the encoding/decoding as a separate component at all. Instead it could be split up into separate pieces, e.g. so that the ACK Manager knows how to do encoding/decoding for ACK frames. Other components know how to do encoding/decoding for their respective frame types. In terms of this PR that might mean moving the ACK encoding/decoding out of quic_wire.c and into quic_ackm.c. We don't have components for the other frame types yet - but we could move the relevant functions as we create them (i.e. defer to later PRs).

File: ssl/quic/quic_wire.c
Comment (mattcaswell) [reply ssl/quic/quic_wire.c:586]: Perhaps a comment explaining some of that?

File: ssl/quic/quic_wire.c
Comment (mattcaswell) [reply ssl/quic/quic_wire.c:619]: Perhaps a comment?

File: ssl/quic/quic_wire.c
Comment (mattcaswell) [reply ssl/quic/quic_wire.c:388]: How do we envisage `ossl_quic_wire_decode_from_ack` and passing a NULL `ack` actually being used? I kind of envisaged calling it once with a NULL `ack`, then allocating the structure with the appropriate ranges and then calling it again with the same `pkt`...is that not how it works? If so we don't really need to advance the packet...although I suppose it would be a bit weird to advance in some cases but not others...

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:42]: I think it would be reasonable to move the ACK structure into this PR and rebase the ACKM PR on top of it. In fact, it's probably a good idea. Wire format encoding is a utility and it makes sense for ACKM to depend on it rather than vice versa.

It would also be possible to have a PR just with struct definitions, and treat the logical representation of the different frame types as something to be committed in itself, depended on by both this and the ACKM PR. We've hit similar issues recently with depending on types for QUIC packet numbers, time representations (which was originally part of Pauli's event queue work but which I also rely on), etc.

Though, where code lives after it's committed is orthogonal to how work is split up into PRs. In practice I think the structures may as well be committed together as part of this PR. That doesn't mean that those structures need to be considered "part" of the wire format encoding functionality. C isn't really modular anyway, all we can really do is decide what header file something lives in.

I think it would be a very bad idea to put ACKM in charge of ACK serialization. These were separated for a reason. The user of the ACKM isn't obliged to send every range described logical ACK provided by the ACKM even, if the number of ranges is high. They can choose what to pack into a packet. The ACKM has high inherent complexity already and moving serialization into it would enhance its complexity and reduce its testability, and the ability for the TXP to take the discretionary decisions it is entitled to when deciding how much space in a packet to allocate to ACK frames.

Comment (mattcaswell) [reply ssl/quic/quic_wire.c:42]: > I think it would be reasonable to move the ACK structure into this PR and rebase the ACKM PR on top of it. In fact, it's probably a good idea. Wire format encoding is a utility and it makes sense for ACKM to depend on it rather than vice versa.

I can live with that.

File: ssl/quic/quic_wire.c
Comment (mattcaswell) [reply ssl/quic/quic_wire.c:388]: Three other possibilities worthy of consideration:

1) Have the `ossl_quic_wire_decode_from_ack` function do the allocation itself

2) Have a separate function that just finds the `total_ranges` value without advancing the PACKET

3) Once `total_ranges` is known the length of the frame can be easily calculated. You could just do `PACKET_forward()` by the calculated amount.

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:398]: Hm, I thought I documented this, but I guess not.

The expectation is that you call `ossl_quic_wire_peek_frame_header` and then call a frame decode function like `ossl_quic_wire_decode_frame_ack` etc., probably via a `switch()` on the frame type or so on.
These functions then re-consume (and advance over) the type field.

(An earlier design had these functions assuming the type field had already been consumed. But the type field contains information which the decode function needs to know in some cases, and getting the caller to pass the type they read felt hacky. This keeps the API clean.)

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:234]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:241]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:252]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:258]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:265]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:275]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:401]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:465]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:473]: !

(plus nit: indentation)

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:483]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:490]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:498]: !

File: include/internal/quic_wire.h
Comment (levitte) [inline include/internal/quic_wire.h:509]: !

**Review by levitte — COMMENTED** [review]
There are a number of places where I'd expect `size_t` to be used, rather than `uint64_t`.  All the maxes, primarly...

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:509]: QUIC uses variable length integers which can be up to 62 bits in size. size_t may not be 64 bits, so we need `uint64_t` here. It should be left to the caller to decide any limits it wants to apply. Values like `max_streams` or `max_data` for example don't have a direct correlation with the limits imposed by the size of the address space so it would be inappropriate for us to limit these values to 32 bits on 32-bit platforms arbitrarily. Fields like `max_data` in particular increase monotonically over the lifetime of a connection and will easily exceed 32 bits with time.

File: include/internal/quic_wire.h
Comment (levitte) [reply include/internal/quic_wire.h:398]: Funny you should mention a switch, 'cause #18838 :wink:

File: include/internal/quic_wire.h
Comment (levitte) [reply include/internal/quic_wire.h:509]: Ah, I hadn't thought of the size aspect (hah!) of `size_t`.  Okie, makes sense, just a bit less expressive from the C standpoint.

File: include/internal/quic_wire.h
Comment (levitte) [reply include/internal/quic_wire.h:398]: (and yeah, I did figure it out from reading `quic_wire.c`, but...

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:388]: You can already do this. If the `num_ack_ranges` in the passed structure is inadequate, as many ranges as possible will be decoded. But you can detect that more were available by examining `*total_ranges`.

Still, a function to get only the number of ranges without advancing the PACKET or decoding the other fields should be slightly cheaper, so I'll add this.

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:98]: Added comment, and renamed to `has_explicit_len` since the current naming is confusing.

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:96]: Hm, you're right. Personally I would support adopting the latter but that's not for this PR. I'll change it.

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:172]: Yes, each peer can have a different exponent it uses to send... but a specific pair of encode and decode calls need to have the same exponent, which is what the comment above is saying.

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:152]: Maybe just `ossl_wire_encode_padding`? I'll change it.

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:226]: It does seem to me that supporting single-copy operation will be very difficult with regards to stream data. I don't think it's an MVP requirement, though?

For now it does occur to me that a function to encode the STREAM/CRYPTO frame headers without allocating the payload bytes may be useful, so I'll add that.

File: include/internal/quic_wire.h
Comment (t8m) [reply include/internal/quic_wire.h:172]: Maybe `This exponent value must match the value used by the peer when decoding.` ?

File: include/internal/quic_wire.h
Comment (t8m) [reply include/internal/quic_wire.h:226]: Yeah, it is definitely not an MVP requirement although it would be nice if we do not close the doors for it in the internal design.

Comment (t8m) [reply include/internal/quic_wire.h:226]: > For now it does occur to me that a function to encode the STREAM/CRYPTO frame headers without allocating the payload bytes may be useful, so I'll add that.

Perfect, that should be sufficient for supporting the single copy if we have a front-end API.

**Comment by hlandau** [comment]
Updated. Rebased, and no longer based on top of ACKM. Some small definitions in ACKM this relied on are now incorporated in this PR. ACKM will be rebased on top of this soon.

I've added testing for malformed frame decoding. Testing of ACK frames with no gaps is already tested.

File: include/internal/quic_wire.h
Comment (t8m) [inline include/internal/quic_wire.h:277]: This comment is slightly misleading as this would be used with single-copy API and the stream data would be pushed by the packetizer to the record layer in a separate buffer.

What about: `To create a well-formed frame, the data from the WPACKET must be immediately followed by f->len bytes of stream data.` Or something similar.

Comment (hlandau) [reply include/internal/quic_wire.h]: Done.

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

**Review by paulidale — COMMENTED** [review]
Looks good.  Mostly nits but a couple of more meaningful comments.

I've not gone through comparing the produced output to the formal frame definitions.

File: include/internal/time.h
Comment (hlandau) [reply include/internal/time.h:19]: This is why I proposed to have `ossl_time_multiply` be `(OSSL_TIME, OSSL_TIME)`. Using different types for coefficients creates more problems than it solves...

In any case, I don't think `OSSL_TIME` constants should ever have a type of `int`. Could make it `uint64_t`.

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:443]: This is confusing to me. I changed the definition of `ossl_time_multiply` so that the second argument is an integer to multiply by (notwithstanding that OSSL_TIME is actually an integer). OSSL_TIME_US is the OSSL_TIME-typed value, the delay value is the multiplier.

File: ssl/quic/quic_wire.c
Comment (hlandau) [reply ssl/quic/quic_wire.c:793]: Any ideas?

File: include/internal/time.h
Comment (paulidale) [reply include/internal/time.h:19]: Integer != int :)  uint64_t is fine.

File: ssl/quic/quic_wire.c
Comment (paulidale) [reply ssl/quic/quic_wire.c:443]: I was under the impression that ack_delay_raw was an OSSL_TIME, if it's not all is good.

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:172]: It would be equally valid. Any preference?

File: ssl/quic/quic_wire.c
Comment (paulidale) [reply ssl/quic/quic_wire.c:443]: I guess:

```c
ossl_time_multiply(OSSL_TIME_US, safe_mul_uint64_t(ack_delay_raw, 1UL << ack_delay_exponent, &err));
if (err)
    ...
```

would be the proper way to do this.

File: ssl/quic/quic_wire.c
Comment (paulidale) [reply ssl/quic/quic_wire.c:793]: Sadly, nothing comes to mind except preprocessor lines :(

Reversing the condition and excluding equality `len_ < SIZE_MAX` sort of circumvents this but is really ugly.

Comment (hlandau) [reply ssl/quic/quic_wire.c:793]: I guess we shouldn't reduce the correctness of the code to satisfy compiler warnings since this is putting the cart before the horse. I'll leave this as it is for now...

If compilers start complaining about this it will definitely happen elsewhere in the codebase too, and it will probably be more productive to figure out a general strategy for it at this time.

File: include/internal/quic_wire.h
Comment (paulidale) [reply include/internal/quic_wire.h:172]: The ID cache proposal now uses fixed arrays.  We might as well be consistent and use them everywhere (and pray the standard doesn't increase the size soon).  Not that I'm against length + pointer.  The reduction in mallocs is probably a good thing.

BTW: I'm currently proposing an array 21 long, with the size in the first byte and the up to 20 octets in the remaining (zero filled).  That way one memcmp is enough to compare two.  It's a bit kludgy but workable.

**Comment by hlandau** [comment]
Updated.

File: include/internal/quic_wire.h
Comment (hlandau) [reply include/internal/quic_wire.h:172]: That's a bit obtuse. Since we will be representing connection IDs all over the place, I'll introduce a structure:
```
#define OSSL_QUIC_MAX_CONN_ID_LEN       20
typedef struct ossl_quic_conn_id_st {
    unsigned char   id_len;
    unsigned char   id[OSSL_QUIC_MAX_CONN_ID_LEN];
} OSSL_QUIC_CONN_ID;

```

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

File: ssl/quic/quic_wire.c
Comment (paulidale) [reply ssl/quic/quic_wire.c:800]: I'm fine with this.  I wasn't expecting a resolution -- I couldn't think of one.
Once we get warnings, we can address them.

We've had some previously BTW which we've managed to avoid.

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged.

**Pull request closed without merging**
